### PR TITLE
Add RAW_RPC_REQUEST and RAW_RPC_RESPONSE to RequestLog and ResponseLog

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

There's currently no way to get the protocol-dependent RPC request and
response objects from RequestLog and ResponseLog.

Modifications:

- Add ApacheThriftCall and ApacheThriftReply
- Add RAW_RPC_REQUEST and RAW_RPC_RESPONSE attribute
- Set RAW_RPC_REQUEST and RAW_RPC_RESPONSE attribute where applicable
- Add test cases that ensures the required attributes are available in
  RequestLog and ResponseLog
- Improve MessageLog.toString()
  - Optimize using StringBuilder in lieu of MoreObjects.ToStringHelper
  - Exclude RAW_RPC_REQUEST and RAW_RPC_RESPONSE from the attributes
  - Exclude 'channel' property since it's often logged by service logger
  - Use simpler attribute names

Result:

- A user can access the low-level details about RPC calls.
- Log messages from LoggingClient/Service look better.

### CODE REVIEW

File: src/main/java/com/linecorp/armeria/common/logging/RequestLog.java [hunk 1]
```diff
@@ -42,6 +49,17 @@
     AttributeKey<RpcRequest> RPC_REQUEST = AttributeKey.valueOf(RequestLog.class, "RPC_REQUEST");
 
     /**
+     * The {@link AttributeKey} of the processed {@link RpcRequest} in its protocol-dependent low-level form.
+     *
+     * <p>For a Thrift request, the value of this {@link Attribute} is an array ({@code Object[]}) whose length
+     * is 2. Its first element is a {@link TMessage} and the second element is a {@link TBase}.
+     *
+     * <p>For a Protocol Buffers request, the value of this {@link Attribute} is a {@link Message} or
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/common/logging/RequestLog.java]: Will this be implemented in a separate PR?

Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/logging/RequestLog.java]: Yes, when I have some time for it..

**Comment by kawamuray** [comment]
General comment about representing attribute value as Object[]:

I'm not sure if an array of object is a good way to gather those values.
However, this attribute would be used rarely so we might don't have to take care of interface that much but if someone want's to use this attribute he/she might have to scan THttpService to understand how this attribute supplied.

For RAW_RPC_REQUEST, it might not present when request contains unknown method name or invalid argument bytes even though server succeeds to deserialize header.
For RAW_RPC_RESPONSE, it doesn't present if a function is oneway, and even though it presents the second element can either be TBase or TApplicationException.
This sounds bit hard to understand this attribute. Users have to write something like below if they want to handle both raw request and raw response for example.

```java
Object reqObjs = req.attr(RequestLog.RAW_RPC_REQUEST).get();
if (reqObjs == null) {
    // request might have contained invalid method name or args
    return;
}

Object[] reqElements = (Object[]) reqObjs;
TMessage reqHeader = (TMessage) reqElements[0];
TBase thriftArgs = (TBase) reqElements[1];

Object resObjs = res.attr(ResponseLog.RAW_RPC_RESPONSE).get();
Object[] resElements = (Object[]) resObjs;

TMessage resHeader = (TMessage) resElements[0];
if (resElements[1] instanceof TApplicationException) {
    appException = (TApplicationException) resElements[1];
} else if (resElements[1] != null) {
    response = (TBase) resElements[1];
} else {
    // one way
}
```

This is bit nsaty I guess; lot's of type casting and magic number indexes. Thing get much worse if the code needs to take care of different protocol(i.e, protobuf).
Can't we have a simple class like below, then user's code can be much cleaner and robust for future breaking change I guess.

```java
public class ThriftMessage {
    private final TMessage header;
    private final TBase<? extends TBase<?, ?>, ? extends TFieldIdEnum> value;
    private final TApplicationException appException;

    public Optional<TBase<...>> getValue() { ... }

    public Optional<TApplicationException> getAppException() {...}
}
```

**Comment by kawamuray** [comment]
How about separating commit for "Fix incorrect 'user-agent' header" this change at least?

File: src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java [hunk 2]
```diff
@@ -209,6 +212,26 @@ private Object decodeResponse(ThriftFunction method, HttpData content) throws TE
                                         result.getClass().getName() + '.' + successField.getFieldName());
     }
 
+    private static TApplicationException readApplicationException(ThriftFunction method,
```
Comment (kawamuray) [inline src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: Let's use same wording for same things. s/method/func/

File: src/main/java/com/linecorp/armeria/server/thrift/THttpService.java [hunk 3]
```diff
@@ -500,7 +500,7 @@ private void decodeAndInvoke(
                 final TApplicationException cause = new TApplicationException(
                         TApplicationException.UNKNOWN_METHOD, "unknown method: " + header.name);
 
-                respond(serializationFormat, seqId, methodName, cause, res);
+                respond(ctx, serializationFormat, seqId, methodName, cause, res);
```
Comment (kawamuray) [inline src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: Kinda question but shouldn't we use `header.name` instead of `methodName` to write response?

File: src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java [hunk 4]
```diff
@@ -178,17 +176,59 @@ private T self() {
...
+            } else {
+                buf.append(name);
+            }
 
-        return helper.add("cause", cause)
-                     .add("attrs", attrs).toString();
+            buf.append('=');
+
+            if (value instanceof Object[]) {
+                buf.append('[');
```
Comment (kawamuray) [inline src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java]: `buf.append(Arrays.toString((Object[]) value));` ?

File: src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java [hunk 5]
```diff
@@ -42,6 +50,18 @@
     AttributeKey<RpcResponse> RPC_RESPONSE = AttributeKey.valueOf(RequestLog.class, "RPC_RESPONSE");
 
     /**
+     * The {@link AttributeKey} of the processed {@link RpcResponse} in its protocol-dependent low-level form.
+     *
+     * <p>For a Thrift response, the value of this {@link Attribute} is an array ({@code Object[]}) whose
+     * length is 2. Its first element is a {@link TMessage} and the second element is a {@link TBase} or
+     * a {@link TException} depending on the type of the {@link TMessage}.
```
Comment (kawamuray) [inline src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java]: I guess TException never appears as second element. Only TApplicationException(it's a TException though :p)?

File: src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java [hunk 6]
```diff
@@ -141,6 +188,43 @@ public void testHttpsInvocation() throws Exception {
...
+        assertTrue(res.hasAttr(ResponseLog.HTTP_HEADERS));
+        assertTrue(res.hasAttr(ResponseLog.RPC_RESPONSE));
+        assertTrue(res.hasAttr(ResponseLog.RAW_RPC_RESPONSE));
+
+        Object[] rawResponse = (Object[]) res.attr(ResponseLog.RAW_RPC_RESPONSE).get();
+        assertThat(rawResponse.length, is(2));
+        assertThat(rawResponse[0], instanceOf(TMessage.class));
+        assertThat(rawResponse[1], instanceOf(TBase.class));
+    }
+
```
Comment (kawamuray) [inline src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java]: How about testing also a case that TApplicationException is provided at least.

**Comment by kawamuray** [comment]
Out of scope but as I just found; can we remove these lines and use ThriftCall.serviceType() instead? https://github.com/line/armeria/blob/master/src/main/java/com/linecorp/armeria/server/thrift/ThriftCallService.java#L96-L105

**Comment by trustin** [comment]
> How about separating commit for "Fix incorrect 'user-agent' header" this change at least?

#284. Please review it as well. :-)

File: src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java [hunk 7]
```diff
@@ -209,6 +212,26 @@ private Object decodeResponse(ThriftFunction method, HttpData content) throws TE
...
+                                                                  TMessage msg) throws TException {
+        TApplicationException appEx;
+        if (msg.type == TMessageType.EXCEPTION) {
+            appEx = TApplicationException.read(inputProtocol);
+            inputProtocol.readMessageEnd();
+        } else if (!method.name().equals(msg.name)) {
+            appEx = new TApplicationException(
+                    TApplicationException.WRONG_METHOD_NAME, msg.name);
+        } else {
+            appEx = null;
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: appEx was meant to be final. Didn't want to assign more than once.

File: src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java [hunk 8]
```diff
@@ -42,6 +50,18 @@
...
+     *
+     * <p>For a Protocol Buffers request, the value of this {@link Attribute} is a {@link Message} or
+     * a {@link MessageLite}.
+     */
+    AttributeKey<Object> RAW_RPC_RESPONSE = AttributeKey.valueOf(RequestLog.class, "RAW_RPC_RESPONSE");
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java]: Yeah, thanks for noticing. :-)

File: src/main/java/com/linecorp/armeria/server/thrift/THttpService.java (hunk 3 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: Couldn't do that because it's the behavior of the official TMultiplexedProtocol implementation.

Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: That being said, let me fix the INVALID_MESSAGE_TYPE case which was using `header.name`.

File: src/main/java/com/linecorp/armeria/server/thrift/THttpService.java [hunk 9]
```diff
@@ -582,48 +584,52 @@ private void invoke(
...
                                 ThriftFunction func, TBase<TBase<?, ?>, TFieldIdEnum> result,
                                 HttpResponseWriter res) {
         respond(serializationFormat,
-                encodeSuccess(serializationFormat, func, seqId, result),
+                encodeSuccess(ctx, serializationFormat, func, seqId, result),
                 res);
     }
 
-    private static void respond(SerializationFormat serializationFormat, int seqId,
+    private static void respond(ServiceRequestContext ctx,
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/thrift/THttpService.java]: Good idea. Let me also replace the `ThriftFunction` parameters in `respond()` methods with `String methodName`.

**Comment by trustin** [comment]
> Out of scope but as I just found; can we remove these lines and use ThriftCall.serviceType() instead? https://github.com/line/armeria/blob/master/src/main/java/com/linecorp/armeria/server/thrift/ThriftCallService.java#L96-L105

Can't do that because the text that comes before the colon may not be a class name.

File: src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java (hunk 6 above)
Comment (trustin) [inline src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java]: Added more tests.

File: src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java (hunk 4 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java]: Done.

File: src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java (hunk 5 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java]: Fixed by introducing `ApacheThriftCall` and `ApacheThriftReply`.

File: src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java (hunk 2 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: Fixed.

**Comment by trustin** [comment]
@kawamuray Ready for another round.

File: src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java [hunk 10]
```diff
@@ -77,16 +98,46 @@ protected String getResponse(String name) {
             sb.sslContext(SessionProtocol.HTTPS, ssc.certificate(), ssc.privateKey());
 
             sb.serviceAt("/hello", THttpService.of(
+                    (AsyncIface) (name, resultHandler) -> resultHandler.onComplete("Hello, " + name + '!')));
+
+            sb.serviceAt("/hellochild", THttpService.of(new HelloServiceChild()));
+
+            sb.serviceAt("/oneway", THttpService.of(
```
Comment (kawamuray) [inline src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java]: isn't used. remove if you prefer to cleanup unused things.

**Comment by kawamuray** [comment]
@trustin  finished second round review and left few comments. Please add fixup commits instead of squashing and force pushing so I could finish reviewing much quicker. I won't merge until you finish squashing commits even I say LGTM :)

File: pom.xml [hunk 11]
```diff
@@ -783,7 +783,10 @@
           </groups>
 
           <links>
-            <link>http://netty.io/4.1/api</link>
+            <link>https://netty.io/4.1/api</link>
+            <link>https://developers.google.com/protocol-buffers/docs/reference/java</link>
+            <link>https://people.apache.org/~thejas/thrift-0.9/javadoc/</link>
+            <link></link>
```
Comment (trustin) [inline pom.xml]: Doh!

File: src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java [hunk 12]
```diff
@@ -42,6 +47,16 @@
     AttributeKey<RpcResponse> RPC_RESPONSE = AttributeKey.valueOf(RequestLog.class, "RPC_RESPONSE");
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/logging/ResponseLog.java]: Thanks! Fixed.

File: src/main/java/com/linecorp/armeria/common/thrift/ApacheThriftCall.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                    "header.type: " + typeStr(header.type) + " (expected: CALL or ONEWAY)");
+        }
+
+        this.args = requireNonNull(args, "args");
+    }
+
+    /**
+     * Returns the arguments of this call.
+     */
+    public TBase<?, ?> args() {
```
Comment (trustin) [inline src/main/java/com/linecorp/armeria/common/thrift/ApacheThriftCall.java]: Because then a user cannot downcast it to the concrete _args or _result implementation:

```java
// compilation error when TBase<? extends TBase<?, ?>, TFieldIdEnum> is returned.
((HelloService.hello_args) call.args()).getName();
```

File: src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java (hunk 10 above)
Comment (trustin) [inline src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java]: Thanks! Fixed.

**Comment by trustin** [comment]
> finished second round review and left few comments. Please add fixup commits instead of squashing and force pushing so I could finish reviewing much quicker. I won't merge until you finish squashing commits even I say LGTM :)

You can click the 'Squash and merge' button, edit the commit message and merge! :-)
https://github.com/blog/2141-squash-your-commits

File: src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java [hunk 14]
```diff
@@ -15,27 +15,44 @@
...
+import com.linecorp.armeria.common.thrift.ThriftReply;
+import com.linecorp.armeria.common.util.Exceptions;
 import com.linecorp.armeria.server.Server;
 import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.logging.LogCollectingService;
 import com.linecorp.armeria.server.logging.LoggingService;
 import com.linecorp.armeria.service.test.thrift.main.HelloService;
 import com.linecorp.armeria.service.test.thrift.main.HelloService.AsyncIface;
+import com.linecorp.armeria.service.test.thrift.main.OnewayHelloService;
```
Comment (kawamuray) [inline src/test/java/com/linecorp/armeria/server/thrift/AbstractThriftOverHttpTest.java]: turns into unused import then

File: src/main/java/com/linecorp/armeria/common/thrift/ApacheThriftCall.java (hunk 13 above)
Comment (kawamuray) [inline src/main/java/com/linecorp/armeria/common/thrift/ApacheThriftCall.java]: ugh

**Comment by kawamuray** [comment]
> You can click the 'Squash and merge' button, edit the commit message and merge! :-)

Didn't know this, great!

**Comment by kawamuray** [comment]
LGTM 👍 Really appreciate you. Will merge after the build succeeds.

**Comment by trustin** [comment]
Thanks for a great review, @kawamuray
